nixos/modules/services/matrix/synapse.nix

```diff
@@ -409,118 +553,31 @@ in {
-                    type = types.bool;
-                    default = true;
-                    example = false;
-                    description = lib.mdDoc ''
-                      Whether to enable TLS on the listener socket.
-                    '';
-                  };
-
-                  x_forwarded = mkOption {
-                    type = types.bool;
-                    default = false;
-                    example = true;
-                    description = lib.mdDoc ''
-                      Use the X-Forwarded-For (XFF) header as the client IP and not the
-                      actual client IP.
-                    '';
-                  };
-
-                  resources = mkOption {
-                    type = types.listOf (types.submodule {
-                      options = {
-                        names = mkOption {
-                          type = types.listOf (types.enum [
-                            "client"
-                            "consent"
-                            "federation"
-                            "keys"
-                            "media"
-                            "metrics"
-                            "openid"
-                            "replication"
-                            "static"
-                          ]);
-                          description = lib.mdDoc ''
-                            List of resources to host on this listener.
-                          '';
-                          example = [
-                            "client"
-                          ];
-                        };
-                        compress = mkOption {
-                          type = types.bool;
-                          description = lib.mdDoc ''
-                            Should synapse compress HTTP responses to clients that support it?
-                            This should be disabled if running synapse behind a load balancer
-                            that can do automatic compression.
-                          '';
-                        };
-                      };
-                    });
-                    description = lib.mdDoc ''
-                      List of HTTP resources to serve on this listener.
-                    '';
-                  };
-                };
-              });
-              default = [ {
+              type = types.listOf listenerType;
+              default = [{
                 port = 8008;
                 bind_addresses = [ "127.0.0.1" ];
                 type = "http";
                 tls = false;
                 x_forwarded = true;
-                resources = [ {
+                resources = [{
                   names = [ "client" ];
                   compress = true;
                 } {
                   names = [ "federation" ];
                   compress = false;
-                } ];
-              } ];
+                }];
+              }] ++ lib.optional hasWorkers {
+                port = 9093;
```

Comment:
There's no real suggested default AFAIK but the worker configuration docs use 9093 in their example: https://matrix-org.github.io/synapse/latest/workers.html#shared-configuration.

If we make it an option we wouldn't be able to move `services.matrix-synapse.workers.config` one level up like you suggested above. It might make sense to document the use of 9093 in the description of `services.matrix-synapse.workers` though.

I think it's technically possible to distribute worker processes over multiple machines but there is no documentation of any kind about this that I could find. Currently, this module doesn't really support worker distribution though because there's no (easy) way to run workers without running the main process as well. I think this could be done as another PR though.

Reply:
I'm fine with that considering that (1) I'd consider it the de-facto standard now considering the docs and (2) it's for internal communication only and we'd need to revise more stuff in this module anyways if we'd ever try to implement multi-machine worker setups (and I'm not sure I want that as long as there's no official documentation on it).

So, please mention port 9093 in e.g. the option's description and we're good.

Reply:
Done.